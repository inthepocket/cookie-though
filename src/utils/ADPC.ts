import { Config, CookiePreferences } from '../types';

export const hasADPC = () => !!navigator.dataProtectionControl;

export const useADPC = async (policies: Config['policies']): Promise<CookiePreferences> => {
  const request = policies.map(policy => ({
    id: policy.category,
    text: `${policy.category}: ${policy.description}`,
  }));

  const { consent } = await navigator.dataProtectionControl!.request(request);

  return {
    isCustomised: true,
    cookieOptions: policies.map(policy => ({
      id: policy.id,
      isEnabled: consent.indexOf(policy.category) !== -1,
    })),
  };
};
