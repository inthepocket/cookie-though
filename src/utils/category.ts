import { Category } from '../types';

export const isEssential = (category: Category) => category === Category.Essential;
