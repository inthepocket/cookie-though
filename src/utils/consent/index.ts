import { Policy } from '../../config';
import { Consent, mapToConsent, mapToPreferences } from '../mappers';
import { getPreferences, mergePreferences, preferencesHaveChanged } from '../preferences';

interface Params {
  policies: Policy[];
  cookiePreferencesKey: string;
}

const getConsent = ({ policies, cookiePreferencesKey }: Params): Consent => {
  const currentPreferences = getPreferences(cookiePreferencesKey);
  const newPreferences = mapToPreferences(policies);
  const defaultConsent = mapToConsent({
    policies,
    preferences: newPreferences,
    isCustomised: false,
  });

  if (!currentPreferences) return defaultConsent;

  if (!preferencesHaveChanged({ newPreferences, currentPreferences })) {
    return mapToConsent({
      policies,
      preferences: currentPreferences,
      isCustomised: true,
    });
  }

  return mapToConsent({
    policies,
    preferences: mergePreferences({ newPreferences, currentPreferences }),
    isCustomised: false,
  });
};

export default getConsent;
