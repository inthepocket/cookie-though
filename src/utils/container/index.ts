import { getFloat } from '../../hooks/useAnimateDetails';
import { getTransitionTime } from '../../hooks/useTransitionDuration';
import { init } from '../../lib';
import inlineCSS from '../css';

export const CONTAINER_CLASS = 'cookie-though';

const createNewContainer = (ariaLabel: string) => {
  const container = document.createElement('aside');
  container.className = CONTAINER_CLASS;
  container.setAttribute('aria-label', ariaLabel);

  const shadowRoot = container.attachShadow({ mode: 'open' });
  const style = document.createElement('style');
  style.textContent = inlineCSS();

  shadowRoot.appendChild(style);

  return { container, shadowRoot };
};

const getContainer = () => {
  return document.querySelector<HTMLElement>(`.${CONTAINER_CLASS}`);
};

const getHiddenContainerBottom = (container: HTMLElement) => {
  const containerHeight = getFloat(container, 'height');

  return `-${containerHeight}px`;
};

const toggleContainer = (isVisible: boolean, bottom?: string) => {
  let container = getContainer();

  if (!container) container = init();

  container.setAttribute('aria-hidden', String(!isVisible));
  container.style.bottom = bottom || getHiddenContainerBottom(container);

  if (!isVisible) {
    setTimeout(
      () => (container!.style.bottom = getHiddenContainerBottom(container!)),
      // Wait for the close transition and only then reset the bottom
      getTransitionTime(container) + 1,
    );
  }
};

export { createNewContainer, getContainer, getHiddenContainerBottom, toggleContainer };
