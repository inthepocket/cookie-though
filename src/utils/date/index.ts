export const getNextYear = () => {
  const now = new Date();

  return new Date(now.setFullYear(now.getFullYear() + 1)).toUTCString();
};
