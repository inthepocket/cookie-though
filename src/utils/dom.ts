export const setVisible = (value: boolean) => {
  const container = document.querySelector<HTMLElement>('.cookie-though')!;
  if (value) {
    container.style.display = 'block';
    container.setAttribute('aria-hidden', 'false');
    return container.classList.add('visible');
  }
  container.setAttribute('aria-hidden', 'true');
  container.classList.remove('visible');
  setTimeout(() => (container.style.display = 'none'), 250);
};
