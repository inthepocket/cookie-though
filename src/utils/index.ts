export * from './category';
export * from './dom';
