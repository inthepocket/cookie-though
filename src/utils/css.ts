import app from 'bundle-text:../components/app.css';
import button from 'bundle-text:../components/button/button.css';
import consent from 'bundle-text:../components/consent/consent.css';
import option from 'bundle-text:../components/consent/option/option.css';
import header from 'bundle-text:../components/header/header.css';
import reset from 'bundle-text:../components/reset.css';

const inlineCSS = () => {
  return reset + app + header + consent + option + button;
};

export default inlineCSS;
