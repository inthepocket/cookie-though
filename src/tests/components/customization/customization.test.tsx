import { h } from 'preact';
import toJson from 'enzyme-to-json';
import { mount, ReactWrapper, shallow } from 'enzyme';
import mockConfig from '../../__mocks__/config';
jest.mock('../../../utils/dom', () => ({
  setVisible: jest.fn(),
}));
import { setVisible } from '../../../utils/dom';

import Button from '../../../components/button';
import Customization from '../../../components/customization';
import ToggleButton from '../../../components/customization/toggleButton';
import Collapse from '../../../components/customization/collapse';
import { isEssential } from '../../../utils';

const defaultProps = {
  cookieOptions: mockConfig.policies.map(policy => ({ ...policy, isEnabled: false })),
  permissionLabels: mockConfig.permissionLabels,
  setCookiePreferences: jest.fn(),
};

describe('Customization', () => {
  afterEach(() => {
    defaultProps.cookieOptions = mockConfig.policies.map(policy => ({
      ...policy,
      isEnabled: false,
    }));
  });
  it('should render properly', () => {
    const wrapper = shallow(<Customization {...defaultProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  const getToggleButton = (wrapper: ReactWrapper) => wrapper.find(ToggleButton).find('button');

  const isCustomizationCollapsed = (wrapper: ReactWrapper) => {
    return !wrapper.find(Collapse).prop('isOpen');
  };

  const getOptionalCookie = (wrapper: ReactWrapper) => wrapper.find('.ct-option').at(1);

  test('clicking the ToggleButton component toggles the isActive state', () => {
    const wrapper = mount(<Customization {...defaultProps} />);
    const acceptButtonLabel = () => {
      return wrapper.find(Button).last().find('button').text();
    };

    getToggleButton(wrapper).simulate('click').update();
    expect(getToggleButton(wrapper).hasClass('ct-active')).toBeTruthy();
    expect(isCustomizationCollapsed(wrapper)).toBeFalsy();
    expect(acceptButtonLabel()).toEqual('Accept');

    getToggleButton(wrapper).simulate('click').update();
    expect(getToggleButton(wrapper).hasClass('ct-active')).toBeFalsy();
    expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
    expect(acceptButtonLabel()).toEqual('Accept all');
  });

  test('clicking the Slider in an Option will toggle the isEnabled state for that option', () => {
    const wrapper = mount(<Customization {...defaultProps} />);
    const getOptionCheckbox = () => {
      return getOptionalCookie(wrapper).find('input').first();
    };

    getOptionCheckbox().simulate('click').update();
    expect(getOptionalCookie(wrapper).hasClass('ct-enabled')).toBeTruthy();

    getOptionCheckbox().simulate('click').update();
    expect(getOptionalCookie(wrapper).hasClass('ct-enabled')).toBeFalsy();
  });

  test('clicking the decline Button will decline all options and save the preferences', () => {
    const enabledCookies = mockConfig.policies.map(mockCookie => ({
      ...mockCookie,
      isEnabled: true,
    }));
    const setCookiePreferences = jest.fn();
    const wrapper = mount(
      <Customization
        {...defaultProps}
        cookieOptions={enabledCookies}
        setCookiePreferences={setCookiePreferences}
      />,
    );
    const declineButton = wrapper.find('button.ct-button-secondary');

    declineButton.simulate('click').update();
    expect(setCookiePreferences).toBeCalledWith({
      isCustomised: true,
      cookieOptions: defaultProps.cookieOptions.map(cookieOption => ({
        id: cookieOption.id,
        isEnabled: isEssential(cookieOption.category) ? true : false,
      })),
    });
    expect(getToggleButton(wrapper).hasClass('ct-active')).toBeFalsy();
    expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
    const optionalCookies = wrapper.find('.option');
    optionalCookies.forEach(optionalCookie => {
      if (!optionalCookie.find('input').prop('disabled')) {
        expect(optionalCookie.hasClass('ct-enabled')).toBeFalsy();
      }
    });
  });

  describe('when clicking the accept button', () => {
    it('will only accept the options a user has enabled and save the preferences', () => {
      const customisedCookies = [
        defaultProps.cookieOptions[0],
        { ...defaultProps.cookieOptions[1], isEnabled: true },
        ...defaultProps.cookieOptions.filter((_, index) => ![0, 1].includes(index)),
      ];
      const setCookiePreferences = jest.fn();
      const wrapper = mount(
        <Customization
          {...defaultProps}
          cookieOptions={customisedCookies}
          setCookiePreferences={setCookiePreferences}
        />,
      );
      const acceptButton = wrapper.find('button').last();

      acceptButton.simulate('click').update();
      expect(setCookiePreferences).toBeCalledWith({
        isCustomised: true,
        cookieOptions: customisedCookies.map(customisedCookie => ({
          id: customisedCookie.id,
          isEnabled: customisedCookie.isEnabled,
        })),
      });
      expect(getToggleButton(wrapper).hasClass('ct-active')).toBeFalsy();
      expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
      expect(getOptionalCookie(wrapper).hasClass('ct-enabled')).toBeTruthy();
      expect(setVisible).toBeCalledWith(false);
    });

    it('will accept all options if the user has none enabled and save the preferences', () => {
      jest.mock('../../../components/app', () => ({
        setVisible: jest.fn(),
      }));
      const setCookiePreferences = jest.fn();
      const wrapper = mount(
        <Customization {...defaultProps} setCookiePreferences={setCookiePreferences} />,
      );
      const acceptButton = wrapper.find('button').last();

      acceptButton.simulate('click').update();
      expect(setCookiePreferences).toBeCalledWith({
        isCustomised: true,
        cookieOptions: defaultProps.cookieOptions.map(cookieOption => ({
          id: cookieOption.id,
          isEnabled: true,
        })),
      });
      expect(getToggleButton(wrapper).hasClass('ct-active')).toBeFalsy();
      expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
      const optionalCookies = wrapper.find('.option').not('.optionSecondary');
      optionalCookies.forEach(optionalCookie => {
        expect(optionalCookie.hasClass('ct-enabled')).toBeTruthy();
      });
      expect(setVisible).toBeCalledWith(false);
    });
  });
});
