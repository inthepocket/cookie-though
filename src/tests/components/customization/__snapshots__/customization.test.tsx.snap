// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customization should render properly 1`] = `
<div>
  <ToggleButton
    isActive={false}
    label="Customize"
    toggleCustomization={[Function]}
  />
  <Collapse
    isOpen={false}
    onWindowResize={[MockFunction]}
  >
    <Options
      isOpen={false}
      onToggle={[Function]}
      options={
        [
          {
            "category": "essential",
            "description": "These cookies are required to run the site.",
            "id": "essential",
            "isEnabled": false,
            "label": "Strictly necessary cookies",
          },
          {
            "category": "functional",
            "description": "We’ll remember the basics such as language.",
            "id": "functional",
            "isEnabled": false,
            "label": "Functional cookies",
          },
          {
            "category": "statistics",
            "description": "We’ll know where we should improve your experience.",
            "id": "statistics",
            "isEnabled": false,
            "label": "Statistics",
          },
          {
            "category": "marketing",
            "description": "We’ll only show you content that interests you.",
            "id": "marketing",
            "isEnabled": false,
            "label": "Marketing",
          },
        ]
      }
    />
  </Collapse>
  <div
    className="ct-acceptance"
  >
    <Button
      className="ct-button-secondary"
      label="Decline"
      onClick={[Function]}
    />
    <Button
      label="Accept all"
      onClick={[Function]}
    />
  </div>
</div>
`;
